### PR TITLE
Converts a CRTF region string to Shape class.

### PR DESCRIPTION
**Pull request by sushobhana**
This is ready to be merged
Todos :
- [x] write docstrings
- [x] Modify shape class to properly hold CRTF regions.
- [x] write documentation
- [x] write better error messages.
- [x] write tests.

### CODE REVIEW

File: regions/io/crtf/read.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+                helper = CRTFRegionParser(self.global_meta, include, type, region_type,
+                                          *crtf_line.group('region', 'parameters'))
+                helper.parse()
+                self.shapes.append(helper.shape)
+            else:
+                self._raise_error("Not a valid CRTF Region type '{0}'.".format(region_type))
+
+        else:
+            self._raise_error("Not a valid CRTF line '{0}'.".format(line))
+            return
```
Comment (keflavich) [inline regions/io/crtf/read.py]: no need for a return after a raise

Comment (keflavich) [reply regions/io/crtf/read.py]: ah, I see, it is necessary because you can set it to 'warn'.  OK.

File: regions/io/crtf/read.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+    def _raise_error(self, msg):
```
Comment (keflavich) [inline regions/io/crtf/read.py]: since this is repeated code, perhaps it belongs in a parent class?

**Comment by sushobhana** [comment]
@keflavich Thanks!! I will change them in the next commit..

File: regions/io/core.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    """
+    def to_region(self):
+        regions = list()
+        for shape in self:
+            # Skip elliptical annulus for now
+            if shape.region_type == 'ellipse' and len(shape.coord) > 5:
+                msg = 'Skipping elliptical annulus {}'.format(shape)
+                warn(msg, DS9RegionParserWarning)
+                continue
+            log.debug(shape)
```
[the remark this answers is not in the document]
Comment (sushobhana) [reply regions/io/core.py]: these were present before. Should I remove all the debug statements?? (There are lot of these in `regions/io/ds9/*` codes)

**Comment by sushobhana** [comment]
W.r.t. the last commit , I have done some changes in the __init__.py of the subpackages, and added some imports in io/core.py . This results to an import error in tests. Can you tell me why is it happening and how can it be resolved? @keflavich

**Comment by sushobhana** [comment]
W.r.t. the last commit , I have done some changes in the init.py of the subpackages, and added some imports in io/core.py . This results to an import error in tests. Can you tell me why is it happening and how can it be resolved? @keflavich

File: regions/io/core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (keflavich) [reply regions/io/core.py]: we can keep them for now, but we should aim to remove them eventually...

**Comment by keflavich** [comment]
@sushobhana, could you add an update to your checklist?  What is the next step for test writing?  Are there parts you need help with?

File: regions/io/crtf/test/data/CRTFgeneral.reg [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+rotbox[[12h01m34.1s, 12d23m33s], [3arcmin, 1arcmin], 12deg], range=[-1240km/s, 1240km/s]
+
+# An annular region, overriding some of the global defaults:
+annulus[[17h51m03.2s, -45d17m50s], [0.10deg, 4.12deg]], corr=[I,Q,U,V], color=red, label='My label here'
+
+# Cuts an ellipse out of the previous regions, but only for Q and a particular frequency range:
+-ellipse[[12deg, +45deg], [0.25deg, 1.34deg], 45rad], range=[1.420GHz, 1.421GHz], corr=[Q], color=green, label='Removed this'
+
+# A diamond marker, in J2000 coordinates:
+symbol[[32.1423deg, 12.1412deg], D], linewidth=2, coord=J2000, symsize=2
```
Comment (sushobhana) [inline regions/io/crtf/test/data/CRTFgeneral.reg]: @keflavich This is the only official sample I am having now. 
Though I can come up with basic test cases of my own.
Like I cannot be sure about the reliability of the Parser until we can have some practical CRTF files. (Like the dec representation as `-32.6.7.8` as we discussed earlier.)

**Comment by sushobhana** [comment]
@keflavich  what do you think about storing meta attribute in the region and checking their validity.(sixth in the checklist) Now that we will also be handling spectral dimensions of the regions, I think we should think of a really good approach . (What about creating a metadict class inheriting from dict ?).  I think this can be moved to a second PR.

**Comment by keflavich** [comment]
I agree that the metadata handling can be moved to another PR.  I like the idea of handling metadata validity checking in a dictionary subclass.

File: regions/io/crtf/test/data/CRTFgeneral.reg (hunk 4 above)
Comment (keflavich) [reply regions/io/crtf/test/data/CRTFgeneral.reg]: I'm starting to search out more examples.  But one straightforward and important thing to do is create round-trip tests to and from ds9 regions.  So, for example, take a ds9 region file, convert each region to its CRTF equivalent, then read it back and convert to ds9 and make sure it's the same.

Comment (sushobhana) [reply regions/io/crtf/test/data/CRTFgeneral.reg]: I still haven't made methods for conversion between CRTF and DS9 . I will make sure I complete that and perform round tests.

File: regions/io/crtf/read.py [hunk 5]
```diff
@@ -29,7 +29,7 @@
 regex_length = re.compile(r'(?:\[[^=]*\])+[,]\s*([\w.+-]*)\]')
 
 # Extracts each 'parameter=value' pair.
-regex_meta = re.compile(r'((?:\w+\s*=\s*[^,\[\]]+)|(?:\w+\s*=\s*\[.*?\]))')
+regex_meta = re.compile(r'(?:(\w+)\s*=[\s\'\"]*([^,\[\]]+?)[\'\",]+)|(?:(\w+)\s*=\s*\[(.*?)\])')
```
Comment (keflavich) [inline regions/io/crtf/read.py]: if at all possible, add a comment describing this regex and/or indicating some strings it should match.  These should also be included in tests, of course, but regex's are really hard to read in general

Comment (sushobhana) [reply regions/io/crtf/read.py]: ya.. I will try to convert them into `VERBOSE` mode (there is a option to describe what you are doing in the regex)

File: regions/io/ds9/read.py [hunk 6]
```diff
@@ -5,42 +5,38 @@
 import re
```
Comment (migueldvb) [inline regions/io/ds9/read.py]: Apart from changing things into `io.core`, are the other updates in `io.ds9` required in this pull request?

File: docs/ds9.rst [hunk 7]
```diff
@@ -17,18 +17,19 @@ representing one DS9 region. The `~regions.Shape` objects can be converted to
     >>> parser = DS9Parser(reg_string)
```
Comment (migueldvb) [inline docs/ds9.rst]: Some commits have `[skip ci]` to bypass running the build. Just a reminder that if this goes in the commit message of the HEAD commit in the pull request the CI build will be passed over for this PR which is not what you want. I learned this by having made that mistake some time ago so better to use [skip ci] only for pull requests that do not need CI.

Comment (migueldvb) [reply docs/ds9.rst]: Perhaps those fixes in ds9 could go in a separate pull request?

**Comment by migueldvb** [comment]
Some tests are failing in Travis CI python 3.6 environment but apparently unrelated with these updates.

File: docs/ds9.rst (hunk 7 above)
Comment (sushobhana) [reply docs/ds9.rst]: Oh! I did not know about that!. I thought it's better to skip ci for minor doc fixes. I will take care of it from next time.
And about those those minor changes in ds9, most of them are related to the changes done in the io/core so I thought it's better to commit in this PR.

File: regions/io/ds9/read.py (hunk 6 above)
Comment (sushobhana) [reply regions/io/ds9/read.py]: Some of them are required, there have been some minor changes in the shape class. There are also some doc fixes which could have gone in a separate PR(My bad!)

**Comment by sushobhana** [comment]
The tests are failing due to some pytest deprecation. Should i be worried? @bsipocz

**Comment by sushobhana** [comment]
@keflavich @migueldvb  It's just that I am making a lot of changes to Shape class  on top of the changes made here. Should I add those changes in this PR?( which is unrelated to converting CRTF strings to shape objects, )..

**Comment by keflavich** [comment]
I'm not sure what you're asking.  Are you saying there are changes critical to this PR that you have not yet included?

I don't know what's going on with the travis failures; we either need to remove the deprecated code (which sounds like an upstream issue) or lock to a slightly older version of pytest.

**Comment by sushobhana** [comment]
No, I have added methods for conversion of regions to shape and also refactored the whole `ds9_objects_to_string` method making them use of the shape layer , on top of the changes done to Shape class in this PR.  https://github.com/sushobhana/regions/tree/improvements.  Should I add those changes here and extend the scope of this PR ?

**Comment by keflavich** [comment]
No, I think it's best to make that a new PR.  This one is nearly ready to merge, right?

**Comment by sushobhana** [comment]
Ya, it's completely ready , just waiting for your signal!

File: regions/io/crtf/test/test_crtf_language.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
+import os
```
Comment (migueldvb) [inline regions/io/crtf/test/test_crtf_language.py]: `os` is imported but unused in the CRFT test.

File: regions/io/crtf/test/test_crtf_language.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+import distutils.version as vers
+
+from numpy.testing import assert_allclose
```
Comment (migueldvb) [inline regions/io/crtf/test/test_crtf_language.py]: `assert_allclose` is apparently unused. Could you check the imports in the test files?

Comment (sushobhana) [reply regions/io/crtf/test/test_crtf_language.py]: sorry , I will remove the unused imports.

**Comment by bsipocz** [comment]
@sushobhana @keflavich - The fix for the pytest related failures is coming with the new releases, ci-helpers will have the update in the next hour or so.

**Pull request merged**